### PR TITLE
dedupe: make largest directory primary

### LINKED ISSUE #2568: rclone dedupe, choose the main folder efficiently
#### What is the problem you are having with rclone?
Let's say 'Folder' and 'Folder (1)' exist in a certain directory, the webUI etc will show them as the same name, just 'Folder', but File Stream etc can't have the same filename so they show as '(1)' etc. Let's say I have 1000 files in 'Folder' and later on 'Folder (1)' is uploaded that has 5 files. I want to run dedupe so that I will end up with just 'Folder' that has 1005 files, or less if some of those 5 files already existed in the original folder. You'd think the way rclone would do this is to treat the 'Folder' as the original one, and copy the contents of 'Folder (1)' into 'Folder' and delete 'Folder (1)'. Instead I see rclone is copying the contents of 'Folder' into 'Folder (1)' and then deleting 'Folder'. This means there's 1000 move (or 'merge' as the CLI shows) operations instead of 5 (takes much longer, more API calls, etc.). It also means the activity pane in the webUI will now show only 'moved x files' and you'll lose the activity of the original 'Folder' because it was deleted.

#### What is your rclone version (eg output from `rclone -V`)
rclone v1.43.1

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Windows 10 64bit

####  Which cloud storage system are you using? (eg Google Drive)
Google Drive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
rclone dedupe GoogleDrive: --dedupe-mode largest -vv

### PR DESCRIPTION
**Pull request by sakshamkhanna**
Fixes  #2568

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by sakshamkhanna** [comment]
Can someone please review the code before I add unit tests

**Comment by sakshamkhanna** [comment]
I accidently force pushed to the fork. I have reverted back, can someone please re-open the pull request.

File: fs/operations/dedupe.go [hunk 1]
```diff
@@ -245,6 +250,70 @@ func dedupeMergeDuplicateDirs(ctx context.Context, f fs.Fs, duplicateDirs [][]fs
...
+		if dirSizes[i] > largestDirSize {
+			largestDirSize = dirSizes[i]
+			largestDirIndex = i
+		}
+	}
+	return
+}
+
+// directorySize returns the size of directory. includes size of subDirectories if recursive == true
+func directorySize(ctx context.Context, f fs.Fs, dir fs.Directory, recursive bool) (size int64, err error) {
```
Comment (ncw) [inline fs/operations/dedupe.go]: I note that there is a function which does pretty much exactly this already.  It would need a dir parameter added.

https://github.com/rclone/rclone/blob/18c029e0f0c3a5dd0c925e285ce39071cb8728ff/fs/operations/operations.go#L1032

File: fs/operations/dedupe.go [hunk 2]
```diff
@@ -245,6 +250,70 @@ func dedupeMergeDuplicateDirs(ctx context.Context, f fs.Fs, duplicateDirs [][]fs
...
+	var dirSize int64
+	entries, err := f.List(ctx, dir.Remote())
```
Comment (ncw) [inline fs/operations/dedupe.go]: However I'm not sure this actually works...

when using `dir.Remote()` you'll get the name of the directory not the ID so rclone will list the same directory each time.

Same with the `Count` function.

**Comment by ivandeex** [comment]
@sakshamkhanna any updates?

**Comment by ivandeex** [comment]
@sakshamkhanna 
Please address review comments after a year.

**Comment by ivandeex** [comment]
@sakshamkhanna
Rebased on master. Fixed build. Fixed linter warnings.

**Comment by ncw** [comment]
I think this PR has a sound idea - making the largest directory primary but the implementation needs fixing

- calling the listing routines on duplicate directories won't work as expected
- listing the directories is expensive - we've already listed them so why not re-use that data

So I think using a simpler approach of finding the directory with the most entries (rather than the largest size) will be more efficient and this can be done by looking through data we already have collected here

https://github.com/rclone/rclone/blob/8a429d12cfb181bf3a28ce4e794b3eada8bf7e0e/fs/operations/dedupe.go#L236-L240

(this could count the items in every directory very easily and make a map of them by ID which gets returned)

**Comment by ivandeex** [comment]
@sakshamkhanna has not been active on github for half a year

**Comment by ivandeex** [comment]
Rebased on current master

**Comment by ivandeex** [comment]
@ncw

>     * listing the directories is expensive - we've already listed them so why not re-use that data
> So I think using a simpler approach of finding the directory with the most entries (rather than the largest size) will be more efficient and this can be done by looking through data we already have collected here

You in fact ask to replace directory-only walk (with `walk.ListDirs` parameter) inside  `dedupeFindDuplicateDirs` by full directory+file walk with `walk.ListAll`. Being in doubt I checked whether some backends which implement `ListR` on the API level can use it to optimize the amount of API traffic. _I would expect that at least Google Drive with its SQL-like queries could support that._
After looking at the `walk.ListR` code (with or without backend ListR support) and especially at the backend [Features.ListR](https://github.com/rclone/rclone/blob/v1.54.0/fs/fs.go#L478) function, which does not imply any API-based filtering, I'm left to admit that your request is absolutely valid.

I'm OK with that, just don't like that after this PR the now all-object walk inside `dedupeFindDuplicateDirs` will be immediately followed by another all-object walk in `Duplicate` (https://github.com/rclone/rclone/blob/v1.54.0/fs/operations/dedupe.go#L354). I don't see a way to avoid that without buffering the complete all-object tree in memory which is wasteful.

File: fs/operations/dedupe.go (hunk 1 above)
Comment (ivandeex) [reply fs/operations/dedupe.go:361]: I think it's impossible to satisfy your `directorySize` request with existing `fs.Fs` interfaces. See below.

File: fs/operations/dedupe.go (hunk 2 above)
Comment (ivandeex) [reply fs/operations/dedupe.go:363]: I think it's impossible to implement `Count` and satisfy your requirements with existing `fs.Fs` interfaces.
See my comment below.

**Comment by ivandeex** [comment]
@ncw 

I created the following directory structure on Google Drive as an experiment:
```
test/dir1
test/dir1/x.txt
test/dir1          (same name, different ID)
test/dir1/y.txt    (in the second one)
```

If I call `fs.Walk` with `ListAll` argument as you instructed, the entries are returned in the following order:
- test/dir1 (first ID)
- test/dir1 (second ID)
- test/dir1/x.txt
- test/dir1/y.txt

We have to find a parent for `x` and `y` to correctly increment directory counters. We cannot use `path.Dir` as both parent paths are equal. Neither generic `Fs.Object` nor concrete implementation in `backend/drive/drive.go` provide a way to get parent ID(s) of an object.
Alternatively we can separately list directories with duplicate names after the first walk one by one to count entries, but this requires extra API calls and still needs a new interface like `Fs.ListByID` which would use directory ID instead of remote path. Neither `List` nor `ListR` allow that now.

In this PR I am adding a new experimental `fs.Object` interface `ParentIDer` with single function `ParentID` that returns parent directory ID of the Object. I implement it in the `drive` and `opendrive` backends. This is possible as drive's `Object` internally keeps `drive.File` in its `baseObject`, which has `Parents` field with the list of parent IDs. You recently wrote somewhere that GDrive does not support multiple parents anymore, so I use just the first one.

I did not add unit test with duplicate directory names because rclone's `Mkdir` and `NewObject` APIs do not allow to fix parent directory ID. Instead I manually created directory structure described above and ran dedupe against it. Additionally I ran integration tests against drive remote.

Please review and comment.
Thanks.

File: fs/operations/dedupe.go (hunk 1 above)
Comment (ivandeex) [reply fs/operations/dedupe.go]: Solved it differently. See my comment below.

File: fs/operations/dedupe.go (hunk 2 above)
Comment (ivandeex) [reply fs/operations/dedupe.go]: Solved it differently. See my comment below.

**Comment by ivandeex** [comment]
I'm not sure whether a secondary optimization in one of subcommands is worth expanding the core internal API. I just don't see any way around it. Leaving for @ncw to decide. If he sees no other utility in new APIs, this pull request can be just closed.

**Comment by ivandeex** [comment]
waiting for review

**Review by ncw — APPROVED** [review]
I think this approach is a good one.

We already added 2 additional methods to make dedupe work so I think adding another isn't bad. It is also a very useful method which I'm sure people would use if we were to expose it in `rclone lsf` and `rclone lsjson`.

I think you've done a good job here and we should merge the code so I'm happy for you to do that when you are ready :-)
